tests/test_update_ssh_keys.rs

```diff
@@ -0,0 +1,622 @@
+                Ok(())
+            }
+            Err(Error(ErrorKind::KeysDisabled(name), _)) => {
+                println!("Skipping add {}, disabled.", name);
+                Ok(())
+            }
+            Err(Error(ErrorKind::KeysExist(_), _)) => {
+                println!("Skipping add {}, already exists.", testname);
+                Ok(())
+            }
+            Err(err) => {
+                add_key_check_results(ssh_dir, aks, keys, pubkeyname, testname, assert_keys);
+                Err(err.into())
+            }
+        }
+    }
+
+    // A wrapper for deleting an ssh key.
+    //
+    // pubkeyname: name of ssh public key, like valid1, bad
+    // testname: key name given as cmdline args, like "one", "two"
+    // assert_keys: Vec of keys to be asserted after add_keys() succeeded.
+    fn del_one_ssh_key(ssh_dir: &str, pubkeyname: &str, testname: &str) {
+        let mut aks = open_authorized_keys(ssh_dir);
+        let keyfiles = [format!("{}/{}.pub", ssh_dir, pubkeyname)];
+
+        let mut keys = vec![];
+        for keyfile in keyfiles {
+            let file = File::open(&keyfile).expect(format!("failed to open keyfile '{:?}'", keyfile).as_str());
+            keys.append(&mut AuthorizedKeys::read_keys(file).unwrap_or_default());
+        }
+
+        for key in aks.remove_keys(testname) {
+            if let AuthorizedKeyEntry::Valid {
+                ref key,
+            } = key
+            {
+                assert!(key.to_fingerprint_string().contains(&*FINGERPRINTS[pubkeyname]));
+
+                // must write before checking if the file exists, because before writing,
+                // authorized_keys.d is not in sync.
+                aks.write().expect("failed to update authorized keys directory");
+
+                let authkeyone: PathBuf = PathBuf::from(format!("{}/authorized_keys.d/{}", ssh_dir, testname));
+
+                assert!(!authkeyone.exists());
+
+                let mut svec: Vec<&str> = Vec::new();
+                svec.push(pubkeyname);
+                assert_has_keys(ssh_dir, svec, true);
+            }
+        }
+
+        // In contrast to write, sync can run only once after the loop.
+        aks.sync().expect("failed to update authorized keys");
+    }
+
+    fn setup_tests() -> TestContext {
+        let ssh_dir_path: PathBuf = TempDir::new(SSH_DIR_PREFIX).unwrap().into_path();
+
+        for (name, text) in &*TEST_KEYS {
+            let pub_path = ssh_dir_path.join(format!("{}.pub", name));
+
+            let mut pubfile = File::create(pub_path.clone()).expect(format!("unable to create a file {}", pub_path.to_str().unwrap_or_default()).as_str());
+            let _ = pubfile.write_all(text.as_bytes());
+        }
+
+        test_no_keys(ssh_dir_path.to_str().unwrap());
+        test_first_run(ssh_dir_path.to_str().unwrap());
+
+        TestContext {
+            ssh_dir: ssh_dir_path,
+        }
+    }
+
+    fn test_no_keys(ssh_dir: &str) {
```

Comment:
```suggestion
    #[test]
    fn test_no_keys() {
    let ctx = setup_tests();
```
This should be an own test